torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -2661,4 +2677,73 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::barrier(
   return work;
 }
 
+void ProcessGroupGloo::monitoredBarrier(const BarrierOptions& opts) {
+  auto monitoredBarrierTimeout = opts.timeout;
+  auto rank = this->getRank();
+  auto t1 = nextTag();
+  auto t2 = nextTag();
+  std::vector<at::Tensor> t = {at::tensor({rank})};
+  // only enforce timeout on rank 0. This is so that other ranks aren't timed
+  // out first, bringing down the job without reporting which rank timed out.
+  if (rank != 0) {
+    auto sendWork = send(t, 0, t1);
+    sendWork->wait();
+    auto recvWork = recv(t, 0, t2);
+    recvWork->wait();
+    return;
+  }
+  auto startTime = std::chrono::steady_clock::now();
+  auto worldSize = this->getSize();
+  for (int rank = 0; rank < worldSize; ++rank) {
```

Comment:
Might be better to have one for loop, where you do send and recv for all ranks and then a second for loop that waits for all the work. That way you don't have this pattern of send-wait for each individual rank.